mautic/mautic
# Support time in date time segment filters

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

Noticed segment filters query builder not respect time parts in query for relative dates (-8 hours,  - 30 minutes)
That means If WE want create segment filter for  datetime field and I will use custom date time field is equal like now -30 hours but less then 8 hours not work properly.
Also this update replace local timezone to UTC, because every where in segment we use toUtcString. Then If the source date is not UTC, there is conflict between times.

Example If today is 13. 1. 2019 9:23:09:

**Relative datetime support for <> operators**

*my filter*

customdatetiemfield  > - 3 hours
and customdatetiemfield  < - 1 hours

*query*:

customdatetiemfield >  '2019-01-13'  and customdatetiemfield  < ' '2019-01-13''

*after PR apply*:

customdatetiemfield > '2019-01-14 06:23:09' and customdatetiemfield  <  '2019-01-14 08:23:09'


**bug: local > UTC**

server timezone: Europe/Amsterdam

filter:

customdatetiemfield  > yesterday

query:

customdatetimefield > '2019-01-13 22:59:59'

expected after bugfix

customdatetimefield > '2019-01-13 23:59:59'


[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Check examples above
2. If you're using mautibox or dev enviroment check your lgos and see sql query of segment. 
3. You should see dates and where statmenets as I describe above

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Test filters from example
3. See If works for you.
